File: _archive/backend-go/src/events/event_bus.go
package events

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
)

// Event types for server lifecycle
const (
	EventServerCreated       = "server.created"
	EventServerUpdated       = "server.updated"
	EventServerDeleted       = "server.deleted"
	EventServerStatusChanged = "server.status.changed"
	EventServerMetrics       = "server.metrics"
	EventServerPlayerJoined  = "server.player.joined"
	EventServerPlayerLeft    = "server.player.left"

	// K8s-specific events (from operator)
	EventK8sResourceCreated = "k8s.resource.created"
	EventK8sResourceUpdated = "k8s.resource.updated"
	EventK8sResourceDeleted = "k8s.resource.deleted"
	EventK8sPodReady        = "k8s.pod.ready"
	EventK8sPodFailed       = "k8s.pod.failed"

	// Sync events
	EventSyncRequired = "sync.required"
	EventSyncComplete = "sync.complete"
)

// ServerEvent represents an event related to a Minecraft server
type ServerEvent struct {
	ID        string                 `json:"id"`
	Type      string                 `json:"type"`
	ServerID  string                 `json:"server_id"`
	TenantID  string                 `json:"tenant_id"`
	Timestamp time.Time              `json:"timestamp"`
	Data      map[string]interface{} `json:"data,omitempty"`
	Source    string                 `json:"source"` // "api", "operator", "metrics"
}

// K8sStateEvent represents K8s state change from operator
type K8sStateEvent struct {
	Type              string    `json:"type"`
	ServerID          string    `json:"server_id"`
	TenantID          string    `json:"tenant_id"`
	Namespace         string    `json:"namespace"`
	ResourceName      string    `json:"resource_name"`
	Phase             string    `json:"phase"`             // Running, Starting, Stopped, Error
	Message           string    `json:"message"`
	ExternalIP        string    `json:"external_ip,omitempty"`
	ExternalPort      int       `json:"external_port,omitempty"`
	PlayerCount       int       `json:"player_count,omitempty"`
	ReadyReplicas     int32     `json:"ready_replicas"`
	DesiredReplicas   int32     `json:"desired_replicas"`
	Timestamp         time.Time `json:"timestamp"`
}

// EventHandler is a function that handles events
type EventHandler func(event *ServerEvent) error

// K8sStateHandler handles K8s state events
type K8sStateHandler func(event *K8sStateEvent) error

// EventBus provides pub/sub messaging using NATS
type EventBus struct {
	conn         *nats.Conn
	js           nats.JetStreamContext
	subscriptions []*nats.Subscription
	handlers     map[string][]EventHandler
	k8sHandlers  []K8sStateHandler
	mutex        sync.RWMutex
	ctx          context.Context
	cancel       context.CancelFunc
}

// EventBusConfig configuration for EventBus
type EventBusConfig struct {
	NATSUrl     string
	ClusterID   string
	ClientID    string
	StreamName  string
	RetryCount  int
	RetryDelay  time.Duration
}

// DefaultEventBusConfig returns default configuration
func DefaultEventBusConfig() *EventBusConfig {
	return &EventBusConfig{
		NATSUrl:    "nats://nats:4222",
		ClusterID:  "minecraft-platform",
		ClientID:   "minecraft-api",
		StreamName: "MINECRAFT_EVENTS",
		RetryCount: 5,
		RetryDelay: 2 * time.Second,
	}
}

// NewEventBus creates a new event bus with NATS connection
func NewEventBus(config *EventBusConfig) (*EventBus, error) {
	if config == nil {
		config = DefaultEventBusConfig()
	}

	ctx, cancel := context.WithCancel(context.Background())

	eb := &EventBus{
		handlers:    make(map[string][]EventHandler),
		k8sHandlers: make([]K8sStateHandler, 0),
		ctx:         ctx,
		cancel:      cancel,
	}

	// Connect to NATS with retry
	var conn *nats.Conn
	var err error

	for i := 0; i < config.RetryCount; i++ {
		conn, err = nats.Connect(config.NATSUrl,
			nats.Name(config.ClientID),
			nats.ReconnectWait(time.Second),
			nats.MaxReconnects(-1),
			nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
				log.Printf("NATS disconnected: %v", err)
			}),
			nats.ReconnectHandler(func(nc *nats.Conn) {
				log.Printf("NATS reconnected to %s", nc.ConnectedUrl())
			}),
			nats.ErrorHandler(func(nc *nats.Conn, sub *nats.Subscription, err error) {
				log.Printf("NATS error: %v", err)
			}),
		)
		if err == nil {
			break
		}
		log.Printf("Failed to connect to NATS (attempt %d/%d): %v", i+1, config.RetryCount, err)
		time.Sleep(config.RetryDelay)
	}

	if err != nil {
		cancel()
		return nil, fmt.Errorf("failed to connect to NATS after %d attempts: %w", config.RetryCount, err)
	}

	eb.conn = conn

	// Create JetStream context for persistent messaging
	js, err := conn.JetStream()
	if err != nil {
		conn.Close()
		cancel()
		return nil, fmt.Errorf("failed to create JetStream context: %w", err)
	}
	eb.js = js

	// Create or update stream for events
	_, err = js.AddStream(&nats.StreamConfig{
		Name:       config.StreamName,
		Subjects:   []string{"server.*", "k8s.*", "sync.*"},
		Retention:  nats.LimitsPolicy,
		MaxAge:     24 * time.Hour,
		MaxMsgs:    100000,
		Discard:    nats.DiscardOld,
		Storage:    nats.FileStorage,
		Replicas:   1,
	})
	if err != nil && err != nats.ErrStreamNameAlreadyInUse {
		log.Printf("Warning: Could not create/update stream: %v", err)
	}

	log.Printf("EventBus connected to NATS at %s", config.NATSUrl)
	return eb, nil
}

// Publish publishes a server event to the event bus
func (eb *EventBus) Publish(event *ServerEvent) error {
	if event.ID == "" {
		event.ID = fmt.Sprintf("%s-%d", event.Type, time.Now().UnixNano())
	}
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now()
	}

	data, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	subject := event.Type
	_, err = eb.js.Publish(subject, data)
	if err != nil {
		return fmt.Errorf("failed to publish event: %w", err)
	}

	log.Printf("Published event: %s (server: %s)", event.Type, event.ServerID)
	return nil
}

// PublishK8sState publishes a K8s state event
func (eb *EventBus) PublishK8sState(event *K8sStateEvent) error {
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now()
	}

	data, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal K8s state event: %w", err)
	}

	subject := fmt.Sprintf("k8s.%s", event.Type)
	_, err = eb.js.Publish(subject, data)
	if err != nil {
		return fmt.Errorf("failed to publish K8s state event: %w", err)
	}

	log.Printf("Published K8s state: %s (server: %s, phase: %s)", event.Type, event.ServerID, event.Phase)
	return nil
}

// Subscribe subscribes to events of a specific type
func (eb *EventBus) Subscribe(eventType string, handler EventHandler) error {
	eb.mutex.Lock()
	defer eb.mutex.Unlock()

	eb.handlers[eventType] = append(eb.handlers[eventType], handler)

	// Create durable consumer for the event type
	sub, err := eb.js.Subscribe(eventType, func(msg *nats.Msg) {
		var event ServerEvent
		if err := json.Unmarshal(msg.Data, &event); err != nil {
			log.Printf("Failed to unmarshal event: %v", err)
			msg.Nak()
			return
		}

		eb.mutex.RLock()
		handlers := eb.handlers[eventType]
		eb.mutex.RUnlock()

		for _, h := range handlers {
			if err := h(&event); err != nil {
				log.Printf("Handler error for %s: %v", eventType, err)
			}
		}

		msg.Ack()
	}, nats.Durable(fmt.Sprintf("api-%s", eventType)), nats.ManualAck())

	if err != nil {
		return fmt.Errorf("failed to subscribe to %s: %w", eventType, err)
	}

	eb.subscriptions = append(eb.subscriptions, sub)
	log.Printf("Subscribed to event type: %s", eventType)
	return nil
}

// SubscribeK8sState subscribes to K8s state events
func (eb *EventBus) SubscribeK8sState(handler K8sStateHandler) error {
	eb.mutex.Lock()
	eb.k8sHandlers = append(eb.k8sHandlers, handler)
	eb.mutex.Unlock()

	// Subscribe to all K8s events
	sub, err := eb.js.Subscribe("k8s.*", func(msg *nats.Msg) {
		var event K8sStateEvent
		if err := json.Unmarshal(msg.Data, &event); err != nil {
			log.Printf("Failed to unmarshal K8s state event: %v", err)
			msg.Nak()
			return
		}

		eb.mutex.RLock()
		handlers := eb.k8sHandlers
		eb.mutex.RUnlock()

		for _, h := range handlers {
			if err := h(&event); err != nil {
				log.Printf("K8s state handler error: %v", err)
			}
		}

		msg.Ack()
	}, nats.Durable("api-k8s-state"), nats.ManualAck())

	if err != nil {
		return fmt.Errorf("failed to subscribe to K8s state events: %w", err)
	}

	eb.subscriptions = append(eb.subscriptions, sub)
	log.Printf("Subscribed to K8s state events")
	return nil
}

// Close closes the event bus connection
func (eb *EventBus) Close() error {
	eb.cancel()

	for _, sub := range eb.subscriptions {
		if err := sub.Unsubscribe(); err != nil {
			log.Printf("Failed to unsubscribe: %v", err)
		}
	}

	eb.conn.Close()
	log.Println("EventBus connection closed")
	return nil
}

// IsConnected returns true if connected to NATS
func (eb *EventBus) IsConnected() bool {
	return eb.conn != nil && eb.conn.IsConnected()
}

// PublishServerCreated publishes a server created event
func (eb *EventBus) PublishServerCreated(serverID, tenantID string, data map[string]interface{}) error {
	return eb.Publish(&ServerEvent{
		Type:      EventServerCreated,
		ServerID:  serverID,
		TenantID:  tenantID,
		Timestamp: time.Now(),
		Data:      data,
		Source:    "api",
	})
}

// PublishServerStatusChanged publishes a server status change event
func (eb *EventBus) PublishServerStatusChanged(serverID, tenantID, oldStatus, newStatus string) error {
	return eb.Publish(&ServerEvent{
		Type:      EventServerStatusChanged,
		ServerID:  serverID,
		TenantID:  tenantID,
		Timestamp: time.Now(),
		Data: map[string]interface{}{
			"old_status": oldStatus,
			"new_status": newStatus,
		},
		Source: "api",
	})
}

// PublishServerDeleted publishes a server deleted event
func (eb *EventBus) PublishServerDeleted(serverID, tenantID string) error {
	return eb.Publish(&ServerEvent{
		Type:      EventServerDeleted,
		ServerID:  serverID,
		TenantID:  tenantID,
		Timestamp: time.Now(),
		Source:    "api",
	})
}

File: api-server/src/events/event-bus.ts
import { EventEmitter } from 'events';
import { v4 as uuidv4 } from 'uuid';

// Event Types
export enum EventType {
  // Server Lifecycle
  SERVER_CREATED = 'server.created',
  SERVER_UPDATED = 'server.updated',
  SERVER_DELETED = 'server.deleted',
  SERVER_STATUS_CHANGED = 'server.status.changed',
  SERVER_METRICS = 'server.metrics',

  // Player Events
  PLAYER_JOINED = 'server.player.joined',
  PLAYER_LEFT = 'server.player.left',

  // Kubernetes Events (from operator)
  K8S_RESOURCE_CREATED = 'k8s.resource.created',
  K8S_RESOURCE_UPDATED = 'k8s.resource.updated',
  K8S_RESOURCE_DELETED = 'k8s.resource.deleted',
  K8S_POD_READY = 'k8s.pod.ready',
  K8S_POD_FAILED = 'k8s.pod.failed',

  // Sync Events
  SYNC_REQUIRED = 'sync.required',
  SYNC_COMPLETE = 'sync.complete',

  // Backup Events
  BACKUP_STARTED = 'backup.started',
  BACKUP_COMPLETED = 'backup.completed',
  BACKUP_FAILED = 'backup.failed',

  // Auth Events
  AUTH_REAUTH_REQUIRED = 'auth.reauth.required',
}

// Base Event Interface
export interface BaseEvent {
  id: string;
  type: EventType;
  timestamp: Date;
  source: 'api' | 'operator' | 'sync' | 'metrics';
  correlationId?: string;
}

// Server Event
export interface ServerEvent extends BaseEvent {
  serverId: string;
  tenantId: string;
  data: Record<string, unknown>;
}

// K8s State Event
export interface K8sStateEvent extends BaseEvent {
  serverId: string;
  tenantId: string;
  namespace: string;
  resourceName: string;
  phase: string;
  message?: string;
  externalIP?: string;
  externalPort?: number;
  playerCount?: number;
  readyReplicas?: number;
  desiredReplicas?: number;
}

// Status Change Event
export interface StatusChangeEvent extends ServerEvent {
  data: {
    oldStatus: string;
    newStatus: string;
    reason?: string;
  };
}

// Event Handler Types
export type EventHandler<T extends BaseEvent = BaseEvent> = (event: T) => void | Promise<void>;

// Event Bus Class
export class EventBus {
  private emitter: EventEmitter;
  private handlers: Map<string, Set<EventHandler<any>>>;

  constructor() {
    this.emitter = new EventEmitter();
    this.emitter.setMaxListeners(100);
    this.handlers = new Map();
  }

  // Publish an event
  publish<T extends BaseEvent>(event: T): void {
    console.log('[EventBus] Publishing %s: %s', event.type, event.id);
    this.emitter.emit(event.type, event);
    this.emitter.emit('*', event); // Wildcard for all events
  }

  // Subscribe to an event type
  subscribe<T extends BaseEvent>(eventType: EventType | '*', handler: EventHandler<T>): () => void {
    const handlers = this.handlers.get(eventType) || new Set();
    handlers.add(handler);
    this.handlers.set(eventType, handlers);

    this.emitter.on(eventType, handler);

    // Return unsubscribe function
    return () => {
      this.emitter.off(eventType, handler);
      handlers.delete(handler);
    };
  }

  // Helper methods for common events

  publishServerCreated(
    serverId: string,
    tenantId: string,
    data: Record<string, unknown> = {}
  ): void {
    this.publish<ServerEvent>({
      id: uuidv4(),
      type: EventType.SERVER_CREATED,
      serverId,
      tenantId,
      timestamp: new Date(),
      source: 'api',
      data,
    });
  }

  publishServerStatusChanged(
    serverId: string,
    tenantId: string,
    oldStatus: string,
    newStatus: string,
    reason?: string
  ): void {
    this.publish<StatusChangeEvent>({
      id: uuidv4(),
      type: EventType.SERVER_STATUS_CHANGED,
      serverId,
      tenantId,
      timestamp: new Date(),
      source: 'api',
      data: { oldStatus, newStatus, reason },
    });
  }

  publishServerDeleted(serverId: string, tenantId: string): void {
    this.publish<ServerEvent>({
      id: uuidv4(),
      type: EventType.SERVER_DELETED,
      serverId,
      tenantId,
      timestamp: new Date(),
      source: 'api',
      data: {},
    });
  }

  publishK8sStateUpdate(event: Omit<K8sStateEvent, 'id' | 'timestamp'>): void {
    this.publish<K8sStateEvent>({
      ...event,
      id: uuidv4(),
      timestamp: new Date(),
    });
  }

  publishSyncRequired(serverId: string, tenantId: string, reason: string): void {
    this.publish<ServerEvent>({
      id: uuidv4(),
      type: EventType.SYNC_REQUIRED,
      serverId,
      tenantId,
      timestamp: new Date(),
      source: 'sync',
      data: { reason },
    });
  }

  publishSyncComplete(serverId: string, tenantId: string): void {
    this.publish<ServerEvent>({
      id: uuidv4(),
      type: EventType.SYNC_COMPLETE,
      serverId,
      tenantId,
      timestamp: new Date(),
      source: 'sync',
      data: {},
    });
  }

  // Get handler count for debugging
  getHandlerCount(eventType: EventType | '*'): number {
    return this.handlers.get(eventType)?.size || 0;
  }
}

// Singleton instance
let eventBusInstance: EventBus | null = null;

export function getEventBus(): EventBus {
  if (!eventBusInstance) {
    eventBusInstance = new EventBus();
  }
  return eventBusInstance;
}
